pyquil/kraus.py

```diff
@@ -56,3 +58,211 @@ def _create_kraus_pragmas(name, qubit_indices, kraus_ops):
+    :param list k2: The list of Kraus operators that are applied first.
+    :return: A combinatorically generated list of composed Kraus operators.
+    """
+    return [np.dot(k1j, k2l) for k1j in k1 for k2l in k2]
+
+
+def damping_after_dephasing(T1, T2, gate_time):
+    """
+    Generate the Kraus map corresponding to the composition
+    of a dephasing channel followed by an amplitude damping channel.
+
+    :param float T1: The amplitude damping time
+    :param float T2: The dephasing time
+    :param float gate_time: The gate duration.
+    :return: A list of Kraus operators.
+    """
+    damping = damping_kraus_map(p=gate_time / T1)
+    dephasing = dephasing_kraus_map(p=gate_time / T2)
+    return combine_kraus_maps(damping, dephasing)
+
+
+# You can only apply gate-noise to non-parametrized gates,
+# so we need to define placeholders for RX(+/- pi/2)
+SINGLE_Q = {
+    "noisy-x-plus90": expm(-1j * np.pi / 4 * np.array([[0, 1],
+                                                       [1, 0]])),
+    "noisy-x-minus90": expm(+1j * np.pi / 4 * np.array([[0, 1],
+                                                        [1, 0]])),
+}
+
+TWO_Q = {
+    "noisy-cz": np.diag([1, 1, 1, -1]),
+}
+
+
+def noisy_instruction(instruction):
+    if isinstance(instruction, Pragma):
+        return instruction
+
+    if instruction.name == 'RZ':
+        return instruction
+
+    if instruction.name == 'I':
+        return instruction
+
+    if instruction.name == 'RX':
+        assert len(instruction.params) == 1
+        assert len(instruction.qubits) == 1
+        param_str = format_parameter(instruction.params[0])
+        if param_str == 'pi/2':
+            return Gate('noisy-x-plus90', [], instruction.qubits)
+        if param_str == '-pi/2':
+            return Gate('noisy-x-minus90', [], instruction.qubits)
+        raise ValueError("Can't add noise to a parametric gate. "
+                         "Try compiling to RX(pi/2) or RX(-pi/2)")
+
+    if instruction.name == 'CZ':
+        return Gate('noisy-cz', [], instruction.qubits)
+
+    raise ValueError('Gate {} is not in the native instruction set'.format(instruction))
+
+
+def _get_program_topology(prog):
+    """
+    Get the graph of two-qubit gates used in a program.
+
+    :param prog: The program
+    :return: (qubits, edges). ``qubits`` is a set of all integer qubits.
+        ``edges`` is a set of tuples of two-qubit pairings.
+    """
+    qubits = set()
+    edges = set()
+    for instruction in prog:
+        if isinstance(instruction, Pragma):
+            continue
+        qb = [q.index for q in instruction.qubits]
+        qubits.update(qb)
+        if len(qb) == 2:
+            edges.add(tuple(sorted(qb)))
+        if len(qb) > 2:
+            raise ValueError("Encountered a >2 qubit instruction")
+
+    return qubits, edges
+
+
+def add_noise_to_program(prog, T1=30e-6, T2=None, gate_time_1q=50e-9, gate_time_2q=150e-09):
```

Comment:
why T1 and not t1?

Reply:
That's how it was in the docs. More than willing to change it if the consensus is pep8 > physics

Reply:
I'm happy to keep it T1, just wanted to make sure it was an explicit choice